### ISSUE TITLE
root_scalar fails when passed a function wrapped with functools.lru_cache

### ISSUE THREAD

**Issue by jbweston** [scipy/scipy#10846]
When you provide `scipy.optimize.root_scalar` with a function wrapped with `functools.lru_cache` it raises a `SystemError`. This bug is also present in previous versions of scipy (I tested back to 0.19) when using the more specific root-finding methods.

#### Reproducing code example:
```
import scipy.optimize
import functools

@functools.lru_cache()
def f(x):
    return x

scipy.optimize.root_scalar(f, bracket=(-1, 1))
```

#### Error message:

```
SystemError: /home/conda/feedstock_root/build_artifacts/python_1562015400360/work/Objects/tupleobject.c:168: bad argument to internal function

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 8, in <module>
    scipy.optimize.root_scalar(f, bracket=(-1, 1))
  File "/home/jbw/.local/miniconda/envs/scipy-test/lib/python3.7/site-packages/scipy-1.4.0.dev0+746aaef-py3.7-linux-x86_64.egg/scipy/optimize/_root_scalar.py", line 249, in root_scalar
    r, sol = methodc(f, a, b, args=args, **kwargs)
  File "/home/jbw/.local/miniconda/envs/scipy-test/lib/python3.7/site-packages/scipy-1.4.0.dev0+746aaef-py3.7-linux-x86_64.egg/scipy/optimize/zeros.py", line 775, in brentq
    r = _zeros._brentq(f, a, b, xtol, rtol, maxiter, args, full_output, disp)
SystemError: <functools._lru_cache_wrapper object at 0x7f48bd1ba930> returned a result with an error set
```

#### Scipy/Numpy/Python version information:
```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
```
```
1.4.0.dev0+746aaef 1.17.2 sys.version_info(major=3, minor=7, micro=3, releaselevel='final', serial=0)
```
but I also observed this problem in scipy `0.19` when using any of the root-finding functions

**Comment by pvanmulbregt** [comment]
Repeated calls with the same arguments to a function should not happen often in the root solving algorithms, so wrapping in a `functools.lru_cache` is unlikely to produce result much sooner for a single run.  But the code should still work.

I looked into it and have a few observations:
1. The problem affects `scipy.optimize.{bisect,brentq,ridder}` and usually `root_scalar` as its default method is `brentq`.
1. The example `f(x)=x` above (and also `f(x)=x**3`) didn't trigger the issue for `scipy.optimize.{newton,toms748}`
1. The Exceptions disappeared if the call to `scipy.optimize.METHOD()` was repeated once or twice more.
1. Changing the endpoints caused the Exceptions to reappear.
1. `scipy.optimize.{bisect,brentq,ridder}` are all written in C.  `scipy.optimize.{newton,toms748}` are written in Python.

I do not know what exactly is the cause, but it does seem that it is a property of the C functions only, not the Python functions.

**Comment by jbweston** [comment]
Ok I've managed to narrow down the problem to [this line](https://github.com/scipy/scipy/blob/master/scipy/optimize/zeros.c#L45):
```
PyArgs(SetItem)(args, 0, Py_BuildValue("d",x));
```
which expands to
```
PyTuple_SetItem(args, 0, , Py_BuildValue("d",x));
```

This returns an error code that we do not catch, which gives rise to the `bad argument to internal function`.

The reason that it returns an error code is because you cannot set an item in a tuple if that tuple is used elsewhere (i.e. it has a refcount > 1).  This is because tuples are meant to be immutable.

lru-cached functions use the `args` tuple as a key in their internal dictionary, which increases the refcount on the args tuple.

TL;DR: we should be creating a new `args` tuple on every call instead of modifying its elements.
